azurerm/internal/services/resource/feature_resource.go

```diff
@@ -0,0 +1,198 @@
+		}
+	}
+
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		return fmt.Errorf("context had no deadline")
+	}
+	stateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{Registering},
+		Target:     []string{Registered},
+		Refresh:    featureRegisteringStateRefreshFunc(ctx, client, id),
+		MinTimeout: 3 * time.Minute,
+		Timeout:    time.Until(deadline),
+	}
+
+	if _, err = stateConf.WaitForStateContext(ctx); err != nil {
+		return fmt.Errorf("waiting for feature(%q) registering to be completed: %+v", id, err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceFeatureRead(d, meta)
+}
+
+func resourceFeatureRead(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Resource.FeaturesClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.FeatureID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ProviderNamespace, id.Name)
+	if err != nil {
+		return fmt.Errorf("error retrieving feature %q: %+v", id, err)
+	}
+	if resp.Properties != nil && resp.Properties.State != nil {
+		if strings.EqualFold(*resp.Properties.State, Pending) {
+			return fmt.Errorf("feature (%q) which requires manual approval can not be managed by terraform", id)
+		}
+		if !strings.EqualFold(*resp.Properties.State, Registered) {
+			return fmt.Errorf("feature (%q) is not registered", id)
+		}
+	}
+
+	d.Set("name", id.Name)
+	d.Set("provider_namespace", id.ProviderNamespace)
+	return nil
+}
+
+func resourceFeatureDelete(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Resource.FeaturesClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.FeatureID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Unregister(ctx, id.ProviderNamespace, id.Name)
+	if err != nil {
+		return fmt.Errorf("error unregistering feature %q: %+v", id, err)
+	}
+
+	if resp.Properties != nil && resp.Properties.State != nil {
+		if strings.EqualFold(*resp.Properties.State, Pending) {
+			return fmt.Errorf("feature (%q) which requires manual approval can not be managed by terraform", id)
+		}
+	}
+
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		return fmt.Errorf("context had no deadline")
+	}
+	stateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{Unregistering},
+		Target:     []string{NotRegistered, Unregistered},
+		Refresh:    featureRegisteringStateRefreshFunc(ctx, client, *id),
+		MinTimeout: 3 * time.Minute,
+		Timeout:    time.Until(deadline),
+	}
+
+	if _, err = stateConf.WaitForStateContext(ctx); err != nil {
+		return fmt.Errorf("waiting for feature(%q) registering to be completed: %+v", id, err)
+	}
+
+	return nil
+}
+
+func featureRegisteringStateRefreshFunc(ctx context.Context, client *features.Client, id parse.FeatureId) pluginsdk.StateRefreshFunc {
```

Comment:
Should this function have an error state for manually approved features to help users? (iirc these features have the state as "Pending" rather than "Registering"?)

Reply:
Yes, I added that at line162, let user know manually approved features can't be managed by tf.